### DISCUSSION TITLE
How do I pass extra data to a collection's <Item>s?

### DISCUSSION THREAD

**Discussion by Vinnl** [adobe/react-spectrum#3891] [Q&A]
When using e.g. [`useTabList`](https://react-spectrum.adobe.com/react-aria/useTabList.html), I'd like every tab to not just have a text label, but also an icon. Looking at the type definitions, I see that e.g. `TabListProps` takes a generic parameter that ends up on `Node`, and allows you to define extra properties (e.g. `Node<{ icon: ReactNode }>`) that you can access when accessing a particular item (e.g. using `props.item.value.icon`).

However, I don't see how I can actually set those properties - I can't just add extra props to `<Item>`, e.g. like this:

```tsx
<Tabs aria-label="Input settings" disabledKeys={['gamepad']}>
  <Item key="mouse" icon={<MouseIcon>}>Mouse Settings</Item>
  <Item key="keyboard" icon={<KeyboardIcon>}>Keyboard Settings</Item>
  <Item key="gamepad" icon={<GamepadIcon>}>Gamepad Settings</Item>
</Tabs>
```

How _do_ you pass custom properties to a collection item? I've seen [this answer](https://github.com/adobe/react-spectrum/discussions/3483#discussioncomment-3579550) to a similar question, but as I understand it that just extends the type definition of an `<Item>` - I don't think that actually allows me to set additional properties that I can access?

(Or alternatively, what's the best way to provide tab icons?)

**Comment by snowystinger** [comment]
Instead of setting it on Item, just send it as a child or as a prop on the child
See https://react-spectrum.adobe.com/react-spectrum/Tabs.html#icons-in-tabs and https://react-spectrum.adobe.com/react-spectrum/Menu.html#complex-menu-items as examples

As for passing other things as props on Item, you can see our implementation of Tab where we allow people to set any data-* attribute they want
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/tabs/src/Tabs.tsx#L145
And then we can access the props here https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/tabs/src/Tabs.tsx#L153
